src/ucp/core/ucp_worker.c

```diff
@@ -2196,18 +2202,27 @@ static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
 static int ucp_worker_discard_remove_filter(const ucs_callbackq_elem_t *elem,
                                             void *arg)
 {
-    return (elem->cb == ucp_worker_discard_uct_ep_progress) ||
-           (elem->cb == ucp_worker_discard_uct_ep_destroy_progress);
+    ucp_worker_h worker;
+    ucp_request_t *req;
+
+    if ((elem->cb == ucp_worker_discard_uct_ep_destroy_progress) ||
+        (elem->cb == ucp_worker_discard_uct_ep_progress)) {
+        req    = (ucp_request_t*)elem->arg;
+        worker = req->send.discard_uct_ep.ucp_worker;
+
+        ucp_request_put(req);
+        ucp_worker_flush_ops_count_dec(worker);
+        return 1;
+    }
+
+    return 0;
 }
 
 static void ucp_worker_discarded_uct_eps_cleanup(ucp_worker_h worker)
 {
     uct_ep_h uct_ep;
-    ucp_request_t *req;
 
-    kh_foreach(&worker->discard_uct_ep_hash, uct_ep, req, {
-        ucp_worker_flush_ops_count_dec(worker);
-        ucp_request_put(req);
+    kh_foreach_key(&worker->discard_uct_ep_hash, uct_ep, {
```

Comment:
comment that ep currently owns the discard operation so ep_destroy will cancel it
BTW what if ep flush already completed, and now we're in the state of ucp_worker_discard_uct_ep_destroy_progress?

Reply:
AFAIU here we are after uct_worker_progress_unregister_safe and ucp_worker_discard_remove_filter, so either we canceled req or it was finished and removed from kh before

Reply:
ok, let's add this to the comment as well